src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,477 @@
+      "Inverting Schur complement for all variables except for 3D points (n = "
+      "%d)",
+      num_params_poses_ + num_params_other_variables_);
+  Eigen::FullPivLU<Eigen::MatrixXd> luOfS_dense(S_dense);
+  if (luOfS_dense.rank() < S_dense.rows()) {
+    LOG(INFO) << StringPrintf(
+        "Unable to compute covariance. The Schur complement on all variables "
+        "(except for 3D points) is rank deficient. Number of columns: %d, "
+        "rank: %d.",
+        S_dense.rows(),
+        luOfS_dense.rank());
+    return false;
+  }
+  cov_variables_ = S_dense.inverse();
+  cov_poses_ = cov_variables_.block(0, 0, num_params_poses_, num_params_poses_);
+  return true;
+}
+
+bool BundleAdjustmentCovarianceEstimator::Compute() {
+  if (!HasValidSchurComplement()) {
+    ComputeSchurComplement();
+  }
+
+  // Schur elimination on other variables to get pose covariance
+  Eigen::MatrixXd S_dense = S_matrix_;
+  LOG(INFO) << StringPrintf("Schur elimination on other variables (n = %d)",
+                            num_params_other_variables_);
+  Eigen::MatrixXd S_aa =
+      S_dense.block(0, 0, num_params_poses_, num_params_poses_);
+  Eigen::MatrixXd S_ab = S_dense.block(
+      0, num_params_poses_, num_params_poses_, num_params_other_variables_);
+  Eigen::MatrixXd S_ba = S_ab.transpose();
+  Eigen::MatrixXd S_bb = S_dense.block(num_params_poses_,
+                                       num_params_poses_,
+                                       num_params_other_variables_,
+                                       num_params_other_variables_);
+  Eigen::LDLT<Eigen::MatrixXd> ldltOfS_bb(S_bb);
+  Eigen::MatrixXd S_poses = S_aa - S_ab * ldltOfS_bb.solve(S_ba);
+
+  // Step 5: Compute covariance
+  LOG(INFO) << StringPrintf(
+      "Inverting Schur complement for pose parameters (n = %d)",
+      num_params_poses_);
+  Eigen::FullPivLU<Eigen::MatrixXd> luOfS_poses(S_poses);
+  if (luOfS_poses.rank() < S_poses.rows()) {
+    LOG(INFO) << StringPrintf(
+        "Unable to compute covariance. The Schur complement on pose parameters "
+        "is rank deficient. Number of columns: %d, rank: %d. This is likely "
+        "due to the poses being underconstrained with Gauge ambiguity.",
+        S_poses.rows(),
+        luOfS_poses.rank());
+    return false;
+  }
+  cov_poses_ = S_poses.inverse();
+  return true;
+}
+
+bool EstimatePoseCovarianceCeresBackend(
+    ceres::Problem* problem,
+    Reconstruction* reconstruction,
+    std::map<image_t, Eigen::MatrixXd>& res) {
+  BundleAdjustmentCovarianceEstimatorCeresBackend estimator(problem,
+                                                            reconstruction);
+  if (!estimator.Compute()) return false;
+  res.clear();
+  for (const auto& image : reconstruction->Images()) {
+    image_t image_id = image.first;
+    if (!estimator.HasPose(image_id)) continue;
+    Eigen::MatrixXd cov = estimator.GetPoseCovariance(image_id);
+    res.insert(std::make_pair(image_id, cov));
```

Comment:
```suggestion
    res.emplace(image_id, estimator.GetPoseCovariance(image_id));
```
... to avoid unnecessary copy of matrix. Same below.